server.go

```diff
@@ -867,39 +867,56 @@ func (s *Server) sendResponse(t transport.ServerTransport, stream *transport.Str
 func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.Stream, srv *service, md *MethodDesc, trInfo *traceInfo) (err error) {
 	if channelz.IsOn() {
 		s.incrCallsStarted()
-		defer func() {
-			if err != nil && err != io.EOF {
-				s.incrCallsFailed()
-			} else {
-				s.incrCallsSucceeded()
-			}
-		}()
 	}
 	sh := s.opts.statsHandler
+	var statsBegin *stats.Begin
 	if sh != nil {
 		beginTime := time.Now()
-		begin := &stats.Begin{
+		statsBegin = &stats.Begin{
 			BeginTime: beginTime,
 		}
-		sh.HandleRPC(stream.Context(), begin)
-		defer func() {
-			end := &stats.End{
-				BeginTime: beginTime,
-				EndTime:   time.Now(),
-			}
-			if err != nil && err != io.EOF {
-				end.Error = toRPCErr(err)
-			}
-			sh.HandleRPC(stream.Context(), end)
-		}()
+		sh.HandleRPC(stream.Context(), statsBegin)
 	}
 	if trInfo != nil {
-		defer trInfo.tr.Finish()
 		trInfo.tr.LazyLog(&trInfo.firstLine, false)
+	}
+	if sh != nil || trInfo != nil || channelz.IsOn() {
```

Comment:
Good call, done. This wouldn't (and doesn't, verified) help with the stack size though as any local variable needs to be allocated stack memory by the compiler at compile-time because there's no way the compiler can know whether it'll needed before runtime.

Reply:
Hmm, I was thinking if the compiler could be sure two local variables had mutually-exclusive scope that it could use the same memory, but maybe it's not that smart.